arm32 support

nathanchance opened the pull request.

nickdesaulniers commented on driver.sh:
```diff
@@ -9,22 +9,50 @@ check_dependencies() {
+      config=multi_v7_defconfig
+      image_name=zImage
+      qemu="qemu-system-arm"
+      qemu_cmdline=( -initrd images/arm/rootfs.cpio
```
It's curious to me that we use a rootfs for aarch64, but an initrd for arm32.  Maybe it doesn't matter, but I was curious if you had strong beliefs on one vs the other @nathanchance ?

nathanchance replied:
Ugh I forgot to try to get an arm32 rootfs working properly, I'll do that tonight. I just pushed what I knew worked.

nathanchance replied:
Alright I got it to work locally, let's see if CI agrees then I can rebase and squash then I think this should be good to go!

nickdesaulniers commented on driver.sh:
```diff
@@ -44,34 +72,29 @@ build_linux() {
 boot_qemu() {
-  local kernel_image=linux/arch/${ARCH}/boot/Image.gz
-  local rootfs=images/${ARCH}/rootfs.ext4
+  local kernel_image=linux/arch/${ARCH}/boot/$image_name
```
mixing `${ARCH}` and `$image_name` in the same statement :-1: Please pick one and be consistent.

nickdesaulniers replied:
I acknowledge I have not been consistent in one vs the other in the script, but mixing in the same statement makes the inconsistency much more obvious.

nathanchance replied:
Hmm what's the issue exactly? They're guaranteed to be set together in `set_variables`.

`$image_name` is also needed for the `make` command in `build_linux` but I guess I can make the `ARCH` a wild card.

nickdesaulniers replied:
Sorry, I meant:

```diff
- local kernel_image=linux/arch/${ARCH}/boot/$image_name
+ local kernel_image=linux/arch/${ARCH}/boot/${image_name}
```

both `ARCH` and `image_name` are variables, right?

nathanchance replied:
OH yes duh... I'll personally use braces on all my variables, I think it makes them stand out a bit but I'll use whatever style you/others prefer, I'm flexible.

nickdesaulniers commented on .travis.yml:
```diff
@@ -26,3 +33,5 @@ before_script:
   - export PATH=${PWD}:${PATH}
   - cd "${OLDPWD}"
+after_script:
+  - sleep 10
```
How come?

nathanchance replied:
When I first tried two jobs, I wasn't getting saved full output until I added that. Looks to be a Travis issue, see the commit message for the link.

nickdesaulniers replied:
Sadly...ok...does cutting this down to `sleep 1` work as reported: https://github.com/travis-ci/travis-ci/issues/6018#issuecomment-312066416 rather than 10?

nathanchance replied:
I'll try, the last commit reference before mine in that issue used 10 so that's what I went with.

nickdesaulniers commented on driver.sh:
```diff
@@ -9,22 +9,51 @@ check_dependencies() {
+      image_name=zImage
+      qemu="qemu-system-arm"
+      qemu_cmdline=( -drive "file=images/arm/rootfs.ext4,format=raw,id=rootfs,if=ide"
+                     -device virtio-blk-device,drive=rootfs
```
What does this `-device` option do? Is it necessary?

groeck commented:
On Fri, Nov 02, 2018 at 11:23:02AM -0700, Nick Desaulniers wrote:
> nickdesaulniers commented on this pull request.
> 
> 
> 
> >    while [[ $# -ge 1 ]]; do
>      case $1 in
>        "-c"|"--clean") cleanup=true ;;
>      esac
>  
>      shift
>    done
> +
> +  # arm64 is the current default if nothing is specified
> +  [[ -z "${ARCH:-}" ]] && ARCH=arm64
> +  export ARCH
> +  case $ARCH in
> +    "arm")
> +      config=multi_v7_defconfig
> +      image_name=zImage
> +      qemu="qemu-system-arm"
> +      qemu_cmdline=( -drive "file=images/arm/rootfs.ext4,format=raw,id=rootfs,if=ide"
> +                     -device virtio-blk-device,drive=rootfs
> 
> What does this `-device` option do? Is it necessary?
> 
Normally it would instantiate virtio-blk and attach the drive with id=rootfs to it,
but then the -drive command should have if=none. Or at least that is how I do
it. Interesting - does this work ?

nathanchance commented:
@groeck Travis says that it does. I never tried `if=none` but I can.

nathanchance commented:
> Cool, thanks for this patch. Would you mind also updating the readme for how to invoke this in order to target arm32 rather than the default arm64?

Done in both fd9b973 and 2e65979.

> but then the -drive command should have if=none.

I've added this change to fd9b973 since it makes sense now that I look into it.

> What does this `-device` option do? Is it necessary?

As Guenter said, it initializes the virtio-blk layer. If I try to initialize the rootfs like the arm64 one, we get a nice fat panic on boot. I'm happy to hear other options but the few qemu arm32 examples I could find used those flags.

```
[   10.157804] VFS: Cannot open root device "vda" or unknown-block(0,0): error -6
[   10.158021] Please append a correct "root=" boot option; here are the available partitions:
[   10.158597] 0100           65536 ram0 
[   10.158620]  (driver?)
[   10.158928] 0101           65536 ram1 
[   10.158935]  (driver?)
[   10.159168] 0102           65536 ram2 
[   10.159174]  (driver?)
[   10.159407] 0103           65536 ram3 
[   10.159413]  (driver?)
[   10.159637] 0104           65536 ram4 
[   10.159643]  (driver?)
[   10.159868] 0105           65536 ram5 
[   10.159874]  (driver?)
[   10.160128] 0106           65536 ram6 
[   10.160134]  (driver?)
[   10.160353] 0107           65536 ram7 
[   10.160360]  (driver?)
[   10.160602] 0108           65536 ram8 
[   10.160608]  (driver?)
[   10.160843] 0109           65536 ram9 
[   10.160849]  (driver?)
[   10.161091] 010a           65536 ram10 
[   10.161099]  (driver?)
[   10.161336] 010b           65536 ram11 
[   10.161342]  (driver?)
[   10.161570] 010c           65536 ram12 
[   10.161576]  (driver?)
[   10.161826] 010d           65536 ram13 
[   10.161832]  (driver?)
[   10.162071] 010e           65536 ram14 
[   10.162077]  (driver?)
[   10.162320] 010f           65536 ram15 
[   10.162327]  (driver?)
[   10.162679] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
[   10.163051] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 4.19.0-cbl-next-20181102-00052-gd5b0c4bc0d40 #1
[   10.163298] Hardware name: Generic DT based system
[   10.163809] [<c0314004>] (unwind_backtrace) from [<c030e484>] (show_stack+0x10/0x14)
[   10.164071] [<c030e484>] (show_stack) from [<c10da6d8>] (dump_stack+0x9c/0xe0)
[   10.164285] [<c10da6d8>] (dump_stack) from [<c0350250>] (panic+0x118/0x2bc)
[   10.164494] [<c0350250>] (panic) from [<c18016cc>] (mount_block_root+0x13c/0x1e8)
[   10.164719] [<c18016cc>] (mount_block_root) from [<c1801da8>] (prepare_namespace+0x1e8/0x1ec)
[   10.164956] [<c1801da8>] (prepare_namespace) from [<c18010a4>] (kernel_init_freeable+0xd8/0xf0)
[   10.165195] [<c18010a4>] (kernel_init_freeable) from [<c10f5524>] (kernel_init+0x8/0x2ac)
[   10.165429] [<c10f5524>] (kernel_init) from [<c03010b4>] (ret_from_fork+0x14/0x20)
[   10.165690] Exception stack(0xdb08dfb0 to 0xdb08dff8)
[   10.165971] dfa0:                                     00000000 00000000 00000000 00000000
[   10.166331] dfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[   10.166650] dfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[   10.167526] ---[ end Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0) ]---
```

Additionally, I've also added the help menu I mentioned earlier in 
2e65979, which should hopefully be clear about how to use the different aspect of the scripts.

6199fd9 is a nice little cleanup so that adding new architectures is even easier.

I personally think this is good to go.

nickdesaulniers commented:
ok, that panic doesn't look clang related; more qemu flag hunting for a later day.  Thanks @nathanchance !
